rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "recorder.hpp"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rosbag2_transport/logging.hpp"
+
+namespace rosbag2_transport
+{
+Recorder::Recorder(std::shared_ptr<rosbag2::Writer> writer, std::shared_ptr<Rosbag2Node> node)
+: writer_(std::move(writer)), node_(std::move(node)) {}
+
+void Recorder::record(const RecordOptions & record_options)
+{
+  auto topics_and_types = record_options.all ?
+    node_->get_all_topics_with_types() :
+    node_->get_topics_with_types(record_options.topics);
+
+  if (topics_and_types.empty()) {
+    throw std::runtime_error("No topics found. Abort");
+  }
+
+  for (const auto & topic_and_type : topics_and_types) {
+    auto topic_name = topic_and_type.first;
+    auto topic_type = topic_and_type.second;
+
+    auto subscription = create_subscription(topic_name, topic_type);
+    if (subscription) {
+      subscriptions_.push_back(subscription);
+      writer_->create_topic({topic_name, topic_type});
+    }
+  }
+
+  if (subscriptions_.empty()) {
+    throw std::runtime_error("No topics could be subscribed. Abort");
+  }
+
+  ROSBAG2_TRANSPORT_LOG_INFO("Subscription setup complete.");
```

Comment:
So I've just seen this change doesn't print the list of recorded topics anymore?
I thought this was just a copy of the existing subscription code?